File: go/string/680_validPalindrome.go
package string

// validPalindrome 最多可以删除一个字符
func validPalindrome(s string) bool {

	// 双指针(左右指针)
	// 如果左右指针对应的值相同，正常移动
	// 否则，判断左右指针之内的字符是否为回文字符串
	l, r := 0, len(s)-1

	for l < r {
		if s[l] == s[r] {
			l++
			r--
		} else {
			flag1, flag2 := true, true
			// 删除r 所在的字符
			for i, j := l, r-1; i < j; i, j = i+1, j-1 {
				if s[i] != s[j] {
					flag1 = false
					break
				}
			}

			// 删除l 所在的字符
			for i, j := l+1, r; i < j; i, j = i+1, j-1 {
				if s[i] != s[j] {
					flag2 = false
					break
				}
			}

			return flag1 || flag2
		}

	}
	return true
}

File: rust/src/string/680_valid_palindrome.rs


pub fn valid_palindrome(s: String) -> bool {
    // 非迭代器解法
    let s: Vec<char> = s.chars().collect::<Vec<char>>();

    let (mut l, mut r) = (0, s.len()-1);

    while l < r {
        if s[l] == s[r] {

            l += 1;
            r -= 1;

        }else {
            let (mut flag1, mut flag2) = (true, true);
            // 删除 r 所在元素
            let (mut i, mut j ) = (l, r-1);
            while i < j {
                if s[i] != s[j] {
                    flag1 = false;
                    break;
                }
                i += 1;
                j -= 1;
            }
            // 删除 l 所在元素
            let (mut i, mut j) = (l+1, r);
            while i < j {
                if s[i] != s[j] {
                    flag2 = false;
                    break;
                }
                i += 1;
                j -= 1;
            }

            return flag1 || flag2
        }
    }

    true
}